### ISSUE TITLE
Bug: Rabbit order can be changed during serialisation

### ISSUE THREAD

**Issue by colonelfazackerley** [andybalaam/rabbit-escape#403]
It was shown in #399 that rabbit order is significant. On android doing something like turning the device 90 degrees forces a serialisation. This can change rabbit order.

**Comment by colonelfazackerley** [comment]
We need a test for serialising at any point in a solution.

**Comment by tttppp** [comment]
*Spoiler* Another problem with the current serialisation is demonstrated by https://github.com/andybalaam/rabbit-escape/issues/107

**Comment by colonelfazackerley** [comment]
I confess I don't understand #107.

Consider the following deist world.
```
  r j  
### ###
#     #
# ### #
#     #
### ###
 A#b#O 
 ##### 
```
The rabbits both die. I have not done this, but I think it would be possible to save them on android. Turn the screen when the rabbits are separated half way down the world. When the world is reloaded, the rabbit order is reversed. The other rabbit gets the bash token, and they escape.

**Comment by andybalaam** [comment]
What do you think is the right answer? Make rabbit order irrelevant, or preserve rabbit order?

**Comment by colonelfazackerley** [comment]
preserve rabbit order. it's not possible to make order irrelevant. it 2 rabbits land in a token cell at the same time, one of them has to pick it up. I would not want to make it random.

i have order serialising and parsing in #401 which is ready for merging.

**Comment by andybalaam** [comment]
If the rabbits are identical it doesn't matter who wins, and if they are not, we can deterministically choose their priority based on their properties. I think this is probably a fruitless line of thought but wanted to float it and see what you thought.

**Comment by colonelfazackerley** [comment]
See my example above. the rabbits may be going in different directions, so it can matter who gets the token.

**Comment by tttppp** [comment]
Fwiw I think persisting rabbit index is the best way to choose between rabbits. We probably shouldn't change this anyway, as it's likely to break some levels!

**Comment by colonelfazackerley** [comment]
Another example of order being important.

```
       
     # 
     # 
rrl  # 
###  # 
### c#O
#######
```

The right rabbit is number 2. Rabbit 2 gets the brolly, then falls slower, so they stack. Rabbit 1 is elder, and so picks up the climb and then dies from the fall the other side of the wall. If the brolly rabbit had got the token, one would have escaped.

**Comment by tttppp** [comment]
That's a really good example. It also suggests situations where a climber and a brollier get to a token from the same direction at the same time. Picking climber over brollier, or vice versa would feel very arbitrary.

**Comment by andybalaam** [comment]
I am convinced. All we want really is an arbitrary but deterministic way of choosing, and order is good enough. As @tttppp points out, other ways seem worse. @GamingInky agrees.

**Comment by colonelfazackerley** [comment]
closed by 77406be1aed40337c00e38d1ee04c15a8fb19c85
